torch/autograd/gradcheck.py

```diff
@@ -661,6 +739,208 @@ def transpose(matrix_of_tensors):
+    # v^T M u = \sum_{i} \sum_{j} u_i * v_j = (\sum_{i} u_i)(\sum_{i} v_i)
+    sum_u = torch.sparse.sum(u) if u.layout == torch.sparse_coo else u.sum()
+    sum_v = torch.sparse.sum(v) if v.layout == torch.sparse_coo else v.sum()
+    return atol * sum_u.item() * sum_v.item()
+
+
+FAST_FAIL_SLOW_OK_MSG = """
+Fast gradcheck failed but element-wise differences are small. This means that the
+test might've passed in slow_mode!
+
+If you are adding a new operator, please file an issue and then use one of the
+workarounds. The workaround depends on how your test invokes gradcheck/gradgradcheck:
+
+If the test
+- manually invokes gradcheck/gradgradcheck, then call gradcheck/gradgradcheck
+  with `fast_mode=False` as a keyword argument.
+- is OpInfo-based (e.g., in test_ops.py), then modify the OpInfo for the test
+  to have `gradcheck_fast_mode=False`
+- is a Module test (e.g., in common_nn.py), then modify the corresponding
+  module_test entry to have `gradcheck_fast_mode=False`
+""".strip()
+
+
+def slow_mode_jacobian_message(func, tupled_inputs, outputs, input_idx, output_idx, rtol, atol):
+    # Compute jacobians in slow mode for better error message
+    slow_numerical = _get_numerical_jacobian(func, tupled_inputs, outputs)[input_idx][output_idx]
+    slow_analytical = _get_analytical_jacobian(tupled_inputs, outputs, input_idx, output_idx)
+
+    # Assume jacobians are non-empty and have the same shape
+    slow_max_diff = (slow_numerical - slow_analytical).abs().max()
+
+    slow_allclose = torch.allclose(slow_analytical, slow_numerical, rtol, atol)
+    msg = ("\nThe above quantities relating the numerical and analytical jacobians are computed \n"
+           "in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background \n"
+           "about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:\n\n"
+           f"Numerical:\n {slow_numerical}\n"
+           f"Analytical:\n{slow_analytical}\n\n"
+           f"The max per-element difference (slow mode) is: {slow_max_diff}.\n")
+    if slow_allclose:
+        # Slow gradcheck would've passed!
+        msg += FAST_FAIL_SLOW_OK_MSG
+    return msg
+
+
+def fast_gradcheck(fail_test, func, func_out, tupled_inputs, outputs, eps, rtol,
+                   atol, raise_exception, check_grad_dtypes, nondet_tol):
+    # Perform the fast version of gradcheck
+    # See https://github.com/pytorch/pytorch/issues/53876 for details
+    inp_tensors = [t for t in tupled_inputs if is_tensor_like(t) and t.requires_grad]
+    inp_tensor_indices = [i for i, t in enumerate(tupled_inputs) if is_tensor_like(t) and t.requires_grad]
+
+    g_cpu = torch.Generator()
+    all_u = [vec_from_tensor(inp, g_cpu) for inp in inp_tensors]
+    all_u_dense = [u.to_dense().reshape(-1) if u.layout == torch.sparse_coo else u for u in all_u]
+    all_v = [vec_from_tensor(out, g_cpu) for out in outputs]
```

Comment:
Its to avoid messing with the user's RNG state. To try to explain it in a hand-wavy way, our `vec_from_tensor` here needs to use random state, so it consume a random number in sequence generated by the default generator or whatever the user is using, so this leads to an inconsistency. I'm not actually sure how the exact mechanics of this works yet, because in theory the test suite should already have some kind of wrapper to reset the RNG state before each invocation or else I don't see how gradcheck would work even in the slow case. Maybe @albanD knows more?

Reply:
@zou3519, Alban looked into this and the culprit is here:
```
@contextlib.contextmanager
def freeze_rng_state():
    rng_state = torch.get_rng_state()
    if torch.cuda.is_available():
        cuda_rng_state = torch.cuda.get_rng_state()
    yield
    if torch.cuda.is_available():
        torch.cuda.set_rng_state(cuda_rng_state)
    torch.set_rng_state(rng_state)
```
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L662
It doesn't set the RNG state as I expected, it actually saves the prior state and restores it after. So this explains the behavior.